ouroboros-consensus/test-storage/Test/Ouroboros/Storage/ImmutableDB/Mock.hs

```diff
@@ -0,0 +1,71 @@
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications    #-}
+module Test.Ouroboros.Storage.ImmutableDB.Mock (openDBMock) where
+
+import           Control.Monad.State (StateT, runStateT)
+import           Data.Proxy
+
+import           Control.Monad.Class.MonadSTM
+
+import           Ouroboros.Consensus.Util ((..:), (.:))
+
+import           Ouroboros.Storage.Common (EpochNo, EpochSize)
+import           Ouroboros.Storage.ImmutableDB.API
+import           Ouroboros.Storage.Util.ErrorHandling (ErrorHandling)
+import qualified Ouroboros.Storage.Util.ErrorHandling as EH
+
+import           Test.Ouroboros.Storage.ImmutableDB.Model
+
+openDBMock  :: forall m hash. (MonadSTM m, Eq hash)
+            => ErrorHandling ImmutableDBError m
+            -> (EpochNo -> EpochSize)
+            -> m (DBModel hash, ImmutableDB hash m)
+openDBMock err epochSize = do
+    dbVar <- atomically $ newTVar dbModel
+    return (dbModel, immDB dbVar)
+  where
+    db :: ImmutableDB hash (StateT (DBModel hash) m)
+    (dbModel, db) = openDBModel err' epochSize
+
+    err' :: ErrorHandling ImmutableDBError (StateT (DBModel hash) m)
+    err' = EH.liftErrState (Proxy @(DBModel hash)) err
+
+    immDB :: TVar m (DBModel hash) -> ImmutableDB hash m
+    immDB dbVar = ImmutableDB
+        { closeDB           = wrap  $   closeDB          db
+        , isOpen            = wrap  $   isOpen           db
+        , reopen            = wrap  .   reopen           db
+        , deleteAfter       = wrap  .   deleteAfter      db
+        , getTip            = wrap  $   getTip           db
+        , getBinaryBlob     = wrap  .   getBinaryBlob    db
+        , getEBB            = wrap  .   getEBB           db
+        , appendBinaryBlob  = wrap  .:  appendBinaryBlob db
+        , appendEBB         = wrap  ..: appendEBB        db
+        , streamBinaryBlobs = wrapI .: streamBinaryBlobs db
+        , immutableDBErr    = err
+        }
+      where
+        wrap  = wrapModel dbVar
+        wrapI = wrapModel dbVar . fmap (wrapIter dbVar)
+
+    wrapModel :: forall a. TVar m (DBModel hash)
```

Comment:
Should this `read`/`write` pair be part of the same transaction?

Reply:
Well-spotted :slightly_smiling_face: 

Ideally, yes, but since we're executing an action of the following type `state -> m state`, we can't.

However, you made me think twice now and I think we should be able to avoid this if we instantiate `m` to `State (DBModel hash)` and do things purely.

Reply:
I have instantiated `m` with `StateT (DBModel hash) (Except ImmutableDBError)` and I was able to get and set the state in a single STM transaction :slightly_smiling_face: